### PR TITLE
Changes to covariance and likelihood calculations

### PR DESCRIPTION
**Pull request by luizfelippesr**
The idea is to add more flexibility to the covariance and likelihood calculations. 

 - [x] Allows user choice of covariance estimator in `Ensemble Likelihood`
 - [x] Implements a diagonal covariance estimator (for testing) and 
 - [x] Implements a diagonal covariance estimator as a  `EnsembleLikelihoodDiagonal` class
 -  ~Precision-matrix-based likelihood calculations (with user choice of precision matrix estimator)~

### CODE REVIEW

File: imagine/likelihoods/ensemble_likelihood.py [hunk 1]
```diff
@@ -60,30 +69,32 @@ def call(self, simulations_dict):
...
+                    sign, logdet = pslogdet(sim_cov*2*np.pi)
+                    likelicache += -0.5*(np.vdot(diff, plu_solve(sim_cov, diff))+sign*logdet)
         else:
             for name in simulations_dict.keys():
-                obs_mean, obs_cov = oas_mcov(simulations_dict[name].data)  # to distributed data
+                sim_mean, sim_cov = self.cov_func(simulations_dict[name].data)  # to distributed data
                 data = deepcopy(self._measurement_dict[name].data)  # to distributed data
-                diff = np.nan_to_num(data - obs_mean)
+                diff = np.nan_to_num(data - sim_mean)
                 if name in self._covariance_dict.keys():  # not all measurements have cov
```
Comment (shutsch) [inline imagine/likelihoods/ensemble_likelihood.py:85]: See comment above

File: imagine/likelihoods/ensemble_likelihood.py [hunk 2]
```diff
@@ -60,30 +69,32 @@ def call(self, simulations_dict):
 
         if self._covariance_dict is None:
```
Comment (shutsch) [inline imagine/likelihoods/ensemble_likelihood.py:70]: I have still some reservations on why we allow for this case. The derivation of our likelihood clearly assumes a existing noise (gaussian) covariance. If we don't know the noise, we should be forced to make explicit assumptions on it. Right now we are implicitly assuming a unit noise standard deviation in whatever units the data has in this case. That means that the results of our inference are not invariant under unit transformations. (e.g. data of  say 5 **cm** would imply a std of 1 **cm**,  changing the  data to 0.05 **m** changes the error to 1 **m**!).

**Review by shutsch — APPROVED** [review]
I agree with all changes, I just left some comments on the no noise covariance case. Maybe I am missing something here, but I think we should discuss this at some point. This is of course not essential for this pull request.

File: imagine/likelihoods/ensemble_likelihood.py (hunk 2 above)
Comment (luizfelippesr) [reply imagine/likelihoods/ensemble_likelihood.py:70]: I agree. It makes little sense to have a measurement without noise. I will remove this case and always require for a covariance dictionary to be provided.

**Comment by luizfelippesr** [comment]
> I agree with all changes, I just left some comments on the no noise covariance case. Maybe I am missing something here, but I think we should discuss this at some point. This is of course not essential for this pull request.

That was a good point. I think the code should simple refuse to work if there are no errors/covariance associated with the data, forcing the user to make explictly their own judgement about the noise when (for whatever reason) this information is unavailable. 

I will remove the  case of `self._covariance_dict==None`.

**Pull request merged**